File: 3. Heap/Heap.java
/*
大顶堆，根比左右节点都大
 */
public class Heap {
    private int[] heap = null;
    private int n = 0;

    public Heap(int k) {
        heap = new int[k];
    }

    /*
    下沉操作
     */
    public void sink(int i) {
        int j = 0;
        int curVal = heap[i];
        while ((j = (i << 1) + 1) < n) {
            if (j < n - 1 && heap[j] < heap[j + 1]) {
                j++;
            }
            if (heap[j] > curVal) {
                heap[i] = heap[j];
                i = j;
            } else {
                break;
            }
        }
        heap[i] = curVal;
    }

    /*
    上浮
     */
    public void swim(int i) {
        int curVal = heap[i];
        int par = 0;

        while (i > 0 && (par = (i - 1) >> 1) != i) {
            if (heap[par] < curVal) {
                heap[i] = heap[par];
                i = par;
            } else {
                break;
            }
        }
        heap[i] = curVal;
    }

    /*
    入堆：
    1. 先把元素加在最后
    2. 进行上浮操作
     */
    public void push(int val) {
        heap[n++] = val;
        swim(n - 1);
    }

    /*
    出堆：
    1. 去除heap[0]的值作为返回值
    2. 将heap[n-1]的存放在heap[0]
    3. 将heap[0]进行下沉操作
     */
    public int pop() {
        int ret = heap[0];
        heap[0] = heap[--n];
        sink(0);
        return ret;
    }

    public int size() {
        return n;
    }
}

File: 03. Heap/heap.py
class Heap:
    def __init__(self):
        self.arr = list()
        self.n = 0

    def _sink(self, i):
        j = 2 * i + 1
        t = self.arr[i]

        while j < self.n:
            if j + 1 < self.n and self.arr[j] < self.arr[j + 1]:
                j += 1

            if self.arr[j] > t:
                self.arr[i] = self.arr[j]
                i = j
                j = 2 * i + 1
            else:
                break

        self.arr[i] = t

    def _swim(self, i):
        t = self.arr[i]
        par = int((i-1)/2)

        while 0 < i != par:

            if self.arr[par] < t:
                self.arr[i] = self.arr[par]
                i = par
                par = int((i - 1) / 1)
            else:
                break
        self.arr[i] = t

    def push(self, val):
        """
        入堆：
        1. 先把元素加在最后
        2. 进行上浮操作
        :param val:
        :return:
        """
        self.arr[self.n] = val
        self.n += 1
        self._swim(self.n-1)

    def pop(self):
        """
        出堆：
        1. 去除heap[0]的值作为返回值
        2. 将heap[n-1]的存放在heap[0]
        3. 将heap[0]进行下沉操作
        :return:
        """
        res = self.arr[0]
        self.n -= 1
        self.arr[0] = self.arr[self.n]
        self._sink(0)
        return res

    def size(self):
        return self.n
